File: 234/Solution_234.java
import java.util.Stack;

class ListNode {
    
    int val;

    ListNode next;

    ListNode(int val){this.val = val;};
}

public class Solution_234 {

    public static boolean isPalindrome(ListNode head){
        
        if (head == null || head.next == null){
            return true;
        }

        Stack<Integer> stack = new Stack<>();
        
        ListNode current = head;

        while ( current != null){
            stack.push(current.val);
            current = current.next;
        }

        current = head;


        while(current != null){
           
           int value = current.val;
           if (value != stack.pop()){
              return false;
           }

          current = current.next;
        }
       return true;

    }

   public static void main(String[] args){
      ListNode head = new ListNode(1);
        head.next = new ListNode(2);
        head.next.next = new ListNode(2);
        head.next.next.next = new ListNode(1); 

        System.out.println(isPalindrome(head));
   }
}

File: 234/Solution_234.cpp
#include <iostream>
#include "ListNode_CPP.h"
#include <stack>

using namespace std;

class Solution {
public:
    bool isPalindromeUseStack(ListNode* head){
        stack<int> st;

        ListNode* p = head;

        while(p!=nullptr){
            st.push(p->val);
            p = p->next;
        }

        p = head;

        while( !p ){
            if (p->val != st.top()){
                return false;
            }
            st.pop();
            p = p->next;
        }

        return true;
    }
};


// 测试代码
int main() {
    // 示例链表: 1 -> 2 -> 2 -> 1
    ListNode* head = new ListNode(1);
    head->next = new ListNode(2);
    head->next->next = new ListNode(2);
    head->next->next->next = new ListNode(1);

    Solution solution;
    bool result = solution.isPalindromeUseStack(head);
    cout << (result ? "true" : "false") << endl; // 输出: true

    // 释放链表内存
    while (head) {
        ListNode* temp = head;
        head = head->next;
        delete temp;
    }

    return 0;
}
